### PR TITLE
Perform excess property checking on intersection and union members

### LINKED ISSUE #13813: No excess properties error in nested intersection type
**TypeScript Version:**  2.1.5

**Code**

```ts
interface A {
    x: string
}

interface B {
    a: A;
}

interface C {
    c: number;
}

type D = B & C;

let a: B = {a: {x: 'hello'}}; // ok
let b: B = {a: {x: 2}}; // error - types of property x are incompatible
let c: B = {a: {x: 'hello', y: 2}}; // error - y does not exist in type A

let d: D = {a: {x: 'hello'}, c: 5}; // ok
let e: D = {a: {x: 2}, c: 5}; // error - types of property x are incompatible
let f: D = {a: {x: 'hello', y: 2}, c: 5}; // should be an error?: y does not exist in type A


```

**Expected behavior:** 
Excess property checking should cause a compiler error for the last example.

**Actual behavior:**
Last example compiles without error.

### LINKED ISSUE #18075: Error "Object literal may only specify known properties" is not happening in all cases with intersection types
**TypeScript Version:**  2.4.2

```ts
export type MyType =
	{
		id: number;
	}
	&
	{
		name: string;
	}
	&
	{
		photo:
		{
			id: number;
		}
		&
		{
			url: string;
		}
	}


export let obj: MyType;

export const photo: typeof obj.photo = {
	id: 1,
	url: '',
	xyz: 1	// Great! This causes an error!
};
	
export const myInstance: MyType = {
	id: 1,
	name: '',
	photo: {
		id: 1,
		url: '',
		xyz: 2	// Bug? This does not cause an error!!!
	}
};
```

**Expected behavior:**

I expect all attributions to `xyz` to give an error.

**Actual behavior:**

TypeScript is accepting one of the attributions to `xyz` (with value 2).

### LINKED ISSUE #28616: Compatibility check for nested conditional types allows unknown properties in second level of nesting
**TypeScript Version:**  3.2.0-dev.20181117

**Code**

```ts
export type View<T> = { [K in keyof T]: T[K] extends object ? boolean | View<T[K]> : boolean };

interface TypeC {
    foo: string;
    bar: string;
}

interface TypeB {
    foo: string,
    bar: TypeC
}

interface TypeA {
    foo: string,
    bar: TypeB,
}

let test: View<TypeA>;

test = { foo: true, bar: true, boo: true } // error TS2322: Type '{ foo: true; bar: true; boo: boolean; }' is not assignable to type 'View<TypeA>'. Object literal may only specify known properties, and 'boo' does not exist in type 'View<TypeA>'.

test = { foo: true, bar: { foo: true, bar: true, boo: true } } // error TS2322:  Type '{ foo: true; bar: { foo: true; bar: true; boo: boolean; }; }' is not assignable to type 'View<TypeA>'. Types of property 'bar' are incompatible. Type '{ foo: true; bar: true; boo: boolean; }' is not assignable to type 'boolean | View<TypeB>'. Object literal may only specify known properties, and 'boo' does not exist in type 'boolean | View<TypeB>'.
```

**Expected behavior:**

```ts
test = { foo: true, bar: { foo: true, bar: { foo: true, bar: true, boo: true } } } // error TS2322:  Type '{ foo: true; bar: { foo: true; bar: { foo: true; bar: true; boo: true }; }; }' is not assignable to type 'View<TypeA>'. Types of property 'bar' are incompatible...
```

**Actual behavior:**

```ts
test = { foo: true, bar: { foo: true, bar: { foo: true, bar: true, boo: true } } } // no error!
```

### PR DESCRIPTION
**Pull request by weswigham**
Fixes #28616
Fixes #18075
Fixes #13813

cc @JasonGore - I believe this enhances excess property checking to the point where it should cover all the faculties JSX tags in 2.8 had.

### CODE REVIEW

**Comment by weswigham** [comment]
@typescript-bot test this & @typescript-bot run dt & @typescript-bot user test this

What a command.

File: src/compiler/checker.ts [hunk 1]
```diff
@@ -7473,6 +7473,13 @@ namespace ts {
             return type.resolvedProperties;
         }
 
+        function getPossiblePropertiesOfUnionType(type: UnionType): Symbol[] {
+            // The following is for effects - getUnionOrIntersectionProperty will cache all the possible union properties into `type`
+            void map(flatMap(type.types, getPropertiesOfType), p => getUnionOrIntersectionProperty(type, p.escapedName));
```
Comment (DanielRosenwasser) [inline src/compiler/checker.ts]: If you're ignoring the result, I would just use `forEach`.

**Comment by RyanCavanaugh** [comment]
RWC Summary
 * One improved error message (we previously failed to elaborate)
 * One change in type readout (differently-ordered union; benign)
 * A somewhat surprising break in VS Code that seems unrelated?

Extracted:
```ts
interface IStringDictionary<V> {
	[name: string]: V;
}
interface INumberDictionary<V> {
	[idx: number]: V;
}

declare function forEach<T>(from: IStringDictionary<T> | INumberDictionary<T>, callback: (entry: { key: any; value: T; }, remove: () => void) => any);

let count = 0;
collections.forEach({ toString: 123 }, () => count++);
                      ~~~~~~~~
!!! error TS2322: Type 'number' is not assignable to type '() => string'.
```
I don't really disagree with the error but it'd be good to understand why it popped up

**Comment by weswigham** [comment]
Yeah, I saw that and was equally confused. Definitely need to look at 1. what that code's even meant to be doing and 2. why does this affect it?

DT results also seem noisy (there's failures on latest master), but there's around 16 new changes there - they all _look_ like new excess prop errors, but it's worth inspecting more closely.

File: src/compiler/checker.ts [hunk 2]
```diff
@@ -12327,11 +12315,24 @@ namespace ts {
...
+                            }
+                        }
                     }
                     else if (target.flags & TypeFlags.Intersection) {
                         isIntersectionConstituent = true; // set here to affect the following trio of checks
-                        result = typeRelatedToEachType(source, target as IntersectionType, reportErrors);
+                        result = typeRelatedToEachType(getRegularTypeOfObjectLiteral(source), target as IntersectionType, reportErrors);
+                        if (result && isPerformingExcessPropertyChecks) {
+                            // Validate against excess props using the original `source`
+                            if (!propertiesRelatedTo(source, target, reportErrors)) {
```
Comment (jack-williams) [inline src/compiler/checker.ts]: Is the `propertiesRelatedTo` check more expensive than the `typeRelatedToEachType` check? Does it make sense to do the checks in the other order?

Comment (weswigham) [reply src/compiler/checker.ts]: I think usually, yeah - we typically consider structural decomposition more expensive than algebraic comparison.

**Comment by weswigham** [comment]
@typescript-bot pack this

**Comment by Jessidhia** [comment]
That error message looks suspiciously like what you'd get from an `extends Object` constraint; or otherwise something that expects your object to conform to the `Object.prototype`.

**Comment by weswigham** [comment]
Wew, so having finally finished going thru them, every related DT break seemed like a really good change where we were catching a bug in either the test code or the type definitions. I've opened:
https://github.com/DefinitelyTyped/DefinitelyTyped/pull/34705
https://github.com/DefinitelyTyped/DefinitelyTyped/pull/34704
https://github.com/DefinitelyTyped/DefinitelyTyped/pull/34702
https://github.com/DefinitelyTyped/DefinitelyTyped/pull/34701
https://github.com/DefinitelyTyped/DefinitelyTyped/pull/34698
https://github.com/DefinitelyTyped/DefinitelyTyped/pull/34697
https://github.com/DefinitelyTyped/DefinitelyTyped/pull/34696
https://github.com/DefinitelyTyped/DefinitelyTyped/pull/34681
https://github.com/DefinitelyTyped/DefinitelyTyped/pull/34680
https://github.com/DefinitelyTyped/DefinitelyTyped/pull/34679
https://github.com/DefinitelyTyped/DefinitelyTyped/pull/34677
https://github.com/DefinitelyTyped/DefinitelyTyped/pull/34676
https://github.com/DefinitelyTyped/DefinitelyTyped/pull/34700
https://github.com/DefinitelyTyped/DefinitelyTyped/pull/34699
https://github.com/DefinitelyTyped/DefinitelyTyped/pull/34671
and 
https://github.com/knockout/knockout/pull/2475

to fix up those breaks on DT - all of them are corrections or improvements to the tests or declarations for a package.

**Comment by weswigham** [comment]
> I don't really disagree with the error but it'd be good to understand why it popped up

Fix't. The `type.propertyCache` is, unfortunately, polluted with any members we find on the apparent type of a thing, rather than just being direct members, so I can't use it verbatim (as apparent type members don't usually come into play during assignability checking). Not really a big deal since we were still iterating over the type to initialize all the properties anyway; it just means I can't sneakily avoid allocating a symbol table to collect unique entries by leveraging an existing one.

File: src/compiler/checker.ts (hunk 2 above)
Comment (jack-williams) [reply src/compiler/checker.ts]: Ah ok, fair enough. I thought that `typeRelatedToEachType` would end up calling into `propertiesRelatedTo` for each union component anyway via `isRelatedTo`.

Comment (weswigham) [reply src/compiler/checker.ts]: It can; but that's the last kind of comparison we make - any other comparison succeeding avoids it.

Comment (jack-williams) [reply src/compiler/checker.ts]: That makes sense. Ty!

File: src/compiler/types.ts [hunk 3]
```diff
@@ -3746,19 +3746,21 @@ namespace ts {
...
-        ReverseMapped     = 1 << 12,        // Property of reverse-inferred homomorphic mapped type
-        OptionalParameter = 1 << 13,        // Optional parameter
-        RestParameter     = 1 << 14,        // Rest parameter
+        ReadPartial       = 1 << 4,         // Synthetic property present in some but not all constituents
+        WritePartial      = 1 << 5,         // Synthetic property present in some but only satisfied by an index signature in others
+        HasNonUniformType = 1 << 6,         // Synthetic property with non-uniform type in constituents
+        HasLiteralType    = 1 << 7,         // Synthetic property with at least one literal type in constituents
+        ContainsPublic    = 1 << 8,         // Synthetic property with public constituent(s)
+        ContainsProtected = 1 << 9,         // Synthetic property with protected constituent(s)
+        ContainsPrivate   = 1 << 10,         // Synthetic property with private constituent(s)
```
Comment (RyanCavanaugh) [inline src/compiler/types.ts]: Off-by-one whitespace 😲

**Comment by weswigham** [comment]
Before I merge: @typescript-bot test this & @typescript-bot run dt & @typescript-bot user test this

**Comment by weswigham** [comment]
Only DT failure is a new copy of a hapi test I already fixed once (they got copied to a namespace while I was working, apparently); opened https://github.com/DefinitelyTyped/DefinitelyTyped/pull/34790, RWC changes are the remaining error message changes; so everything looks good. Going to merge~

**Pull request merged**